mp_cnn/dataset.py

```diff
@@ -104,12 +110,35 @@ def initialize(self, word_index, embedding):
             sent_a_tokens.append(sa_tokens)
             sent_b_tokens.append(sb_tokens)
 
+            unique_tokens = set(sa_tokens) | set(sb_tokens)
+            for t in unique_tokens:
+                word_to_doc_cnt[t] += 1
+
         self.sentences = []
+        stoplist = set(stopwords.words('english'))
         for i in range(len(sent_a)):
             sent_pair = {}
             sent_pair['a'] = self._get_sentence_embeddings(sent_a_tokens[i], word_index, embedding)
             sent_pair['b'] = self._get_sentence_embeddings(sent_b_tokens[i], word_index, embedding)
+
+            tokens_a_set, tokens_b_set = set(sent_a_tokens[i]), set(sent_b_tokens[i])
+            intersect = tokens_a_set & tokens_b_set
+            overlap = len(intersect) / (len(tokens_a_set) + len(tokens_b_set))
+            df_overlap = sum(word_to_doc_cnt[w] for w in intersect)
+            df_weighted_overlap = df_overlap / (len(tokens_a_set) + len(tokens_b_set))
+
+            tokens_a_set_no_stop = set(w for w in sent_a_tokens[i] if w not in stoplist)
+            tokens_b_set_no_stop = set(w for w in sent_b_tokens[i] if w not in stoplist)
+            intersect_no_stop = tokens_a_set_no_stop & tokens_b_set_no_stop
+            overlap_no_stop = len(intersect_no_stop) / (len(tokens_a_set_no_stop) + len(tokens_b_set_no_stop))
+            df_no_stop = sum(word_to_doc_cnt[w] for w in intersect_no_stop)
+            df_weighted_overlap_no_stop = df_no_stop / (len(tokens_a_set_no_stop) + len(tokens_b_set_no_stop))
```

Comment:
do you mean to weight this with the IDF scores?

Reply:
I'm trying to be consistent with our SM model code. In our SM-CNN-Torch repo (https://github.com/castorini/SM-CNN-Torch/blob/aebdb8f4b7086fbbfea0e69680ea2331a1697961/scripts/overlap_features.py#L45) it is doing this:

```python
for w in word_overlap:
    df_overlap += word2df[w]
df_overlap /= (len(q_set) + len(a_set))
```

Which is I'm doing here. You're right that this is not inverse document frequency - but it is a proxy for idf?

Reply:
`df_weighted_overlap_no_stop` in your code weights the overlap by the document frequency; it should be IDF since rarer words get higher importance. The original code by the authors had bugs. You can do something like this: https://github.com/castorini/data/blob/master/TrecQA/overlap_features.py#L112

Reply:
Updated. @rosequ please review again :)